src/main/java/frc/robot/subsystems/drive/Drive.java

```diff
@@ -0,0 +1,227 @@
+
+        SwerveModule fl = new SwerveModule(-Constants.Drive.BASE_WIDTH / 2,  Constants.Drive.BASE_LENGTH / 2, 37.5, 0.0, 1, 2, 0);// FIXME
+        SwerveModule fr = new SwerveModule( Constants.Drive.BASE_WIDTH / 2,  Constants.Drive.BASE_LENGTH / 2, -150.2, 0.0, 5, 6, 3);
+        SwerveModule bl = new SwerveModule(-Constants.Drive.BASE_WIDTH / 2, -Constants.Drive.BASE_LENGTH / 2, -158.2, 180.0, 3, 4, 1);
+        SwerveModule br = new SwerveModule( Constants.Drive.BASE_WIDTH / 2, -Constants.Drive.BASE_LENGTH / 2, -53.3, 180.0, 7, 8, 2);
+
+        _swerveModules = new SwerveModule[]// turn the swerve modules into an array, so they can be easily accessed, with an arbitrary number of them
+        { 
+            fl, 
+            fr,
+            bl,
+            br 
+        };
+
+        setOrigin(0, 0);// rotate around the center of the robot, by default
+        resetOdometer();// initialize the odometer to 0, 0
+    }
+
+    public void chassisDrive(double drive, double strafe, double rotate)
+    {
+        Vector translateVector = new Vector(strafe, drive);
+        
+        drive(translateVector, rotate);
+    }
+
+    public void fieldDrive(double x, double y, double rotate)
+    {
+        Vector translateVector = new Vector(x, y);
+
+        translateVector.translatePolarPosition(0, -getHeading());
+        
+        drive(translateVector, rotate);
+    }
+
+    private void drive(Vector translateVector, double rotate)
+    {
+        Vector[] moduleCommands = new Vector[_swerveModules.length];
+
+        double maxSpeed = 0;
+
+        for (int i = 0; i < _swerveModules.length; i++)
+        {
+            Vector modulePosition = _swerveModules[i].subtract(_origin);
+            Vector rotateVector = new Vector(modulePosition.getY(), -modulePosition.getX());// clockwise 90 deg.
+            rotateVector = rotateVector.multiply(rotate / Constants.Drive.TYPICAL_MODULE_DIST);
+            Vector outputVector = translateVector.add(rotateVector);
+
+            moduleCommands[i] = outputVector;
+
+            maxSpeed = Math.max(maxSpeed, moduleCommands[i].getMagnitude());
+        }
+
+        for (int i = 0; i < moduleCommands.length; i++)
+        {
+            if (maxSpeed > 1)
+            {
+                moduleCommands[i] = moduleCommands[i].divide(maxSpeed);
+            }
+
+            _swerveModules[i].drive(moduleCommands[i]);
+        }
+    }
+
+    public void zeroModuleRotations()
+    {
+        for (int i = 0; i < _swerveModules.length; i++)
+        {
+            _swerveModules[i].zeroEncoder();
+        }
+    }
+
+    public void setOrigin(double x, double y)
+    {
+        setOrigin(new Vector(x, y));   
+    }
+
+    public void setOrigin(Vector newOrigin)
+    {
+        _origin = newOrigin;
+
+        for (int i = 0; i < _swerveModules.length; i++)
+        {
+            Vector modulePosition = _swerveModules[i].clone();
+            modulePosition = modulePosition.subtract(_origin);
```

Comment:
I don't think this does what you want it to do. We create a new instance of a swerve module and change its position relative to the new origin, but we never update the variable in `_swerveModules[i]`. The old swerve module is never updated (because we cloned it).